src/exo/range_analysis.py

```diff
@@ -1,104 +1,171 @@
+from __future__ import annotations
 from collections import ChainMap
+from dataclasses import dataclass
+from typing import Optional
 
-from .LoopIR import LoopIR, T
+from .LoopIR import LoopIR, T, LoopIR_Compare, get_reads_of_expr
 from .new_eff import Check_ExprBound, Check_ExprBound_Options
 
 
-def index_range_analysis(expr, env):
+def binop(op: str, e1, e2):
+    return LoopIR.BinOp(op, e1, e2, e1.type, e1.srcinfo)
+
+
+@dataclass
+class IndexRange:
     """
-    Performs range-analysis on an index expression.
-
-    Function takes in an environment in which the expression exists.
-    The environment is a mapping from `Sym` -> range.
-    It calculates the range of possible values that the index
-    expression could be in.
-
-    Any range, in the environment mapping or the result of the
-    analysis is either:
-        A tuple `T` of length 2 s.t. it represents the range
-        `[T[0], T[1]]` (both inclusive).
-        If `T[0]` or `T[1]` is `None` it represents no knowledge
-        of the value of that side of the range.
+    Represents a range of possible values between [base + lo, base + hi]
+        - [base] contains all expressions that we don't want to bounds infer (e.g. free
+        variables). If base is None, it signifies a base of 0.
+        - [lo] and [hi] are bounds. If either is None, it means that there is no
+        constant bound
```

Comment:
I made the mistake of starting the convention of using None as an unkown. I think it might be a good idea to implement a type a that is basically Integers+Unknwon and hide all the messiness behind the implementation of that type's operator. I can work on it tomorrow.

Reply:
Isn't that essentially what I have with `Optional[int]`?

Reply:
Yes, but I am thinking we can then do the ranges merging without having to do all the is None checks

Reply:
I think we can probably just change that later? I think that change is fairly orthogonal to everything we need for the paper deadline, and may possibly be irrelevant if we ultimately replace some of this with abstract interpretation.